lib/typecheck/pretypecheck.ml

```diff
@@ -6,94 +6,102 @@ open Common
 
-    let message_arity_error tag expected_len actual_len =
+    let message_arity_error pos tag expected_len actual_len =
         let msg =
             asprintf "Arity error. Message '%s' expects %d arguments, but %d were provided."
-            tag expected_len actual_len
+                tag expected_len actual_len
         in
-        raise (pretype_error msg)
+        raise (pretype_error msg [pos])
 
-    let unbound_variable var =
+    let unbound_variable pos var =
         let msg =
-            asprintf "Unbound variable %s" var
+            asprintf "Unbound variable %s." var
         in
-        raise (pretype_error msg)
+        raise (pretype_error msg [pos])
 
-    let type_mismatch expected actual =
+    let type_mismatch pos_list expected actual =
         let msg =
-            asprintf "Type mismatch. Expected %a but got %a."
+            asprintf "Type mismatch. Expected %a but got %a."                
                 Pretype.pp expected
                 Pretype.pp actual
         in
-        raise (pretype_error msg)
+        raise (pretype_error msg pos_list)
 
-    let _type_mismatch_with_message msg expected actual =
+    (* let type_mismatch_with_message pos custom_msg expected actual =
         let msg =
             asprintf "%s. Expected %a but got %a."
-                msg
+                custom_msg
                 Pretype.pp expected
                 Pretype.pp actual
         in
-        raise (pretype_error msg)
+        raise (pretype_error msg) *)
 
-    let type_mismatch_with_expected msg actual =
+    let type_mismatch_with_expected pos expected_msg actual =
         let msg =
-            asprintf
-                "Type mismatch. Expected %s but got %a."
-                msg
+            asprintf "Type mismatch. Expected %s but got %a."                
+                expected_msg
                 Pretype.pp actual
         in
-        raise (pretype_error msg)
+        raise (pretype_error msg [pos])
 
-    let _empty_guards () =
-        raise (pretype_error "Guards cannot be empty")
+    (* let empty_guards pos () =
+        let msg =
+            asprintf "Guards cannot be empty \n At %a" Position.pp pos
+        in
+        raise (pretype_error msg) *)
 
-    let cannot_synth_empty_guards () =
-        raise (pretype_error "Need at least one non-fail guard to synthesise the type for a 'guard' expression.")
+    let cannot_synth_empty_guards pos () =
+        let msg =
+            asprintf "Need at least one non-fail guard to synthesise the type for a 'guard' expression." 
+        in
+        raise (pretype_error msg [pos])
 
-    let cannot_synth_fail () =
-        raise (pretype_error "Cannot synthesise a type for a 'fail' guard")
+    let cannot_synth_fail pos () =
+        let msg =
+            asprintf "Cannot synthesise a type for a 'fail' guard."
+        in
+        raise (pretype_error msg [pos])
 
-    let cannot_synth_sum term =
+    let cannot_synth_sum pos term =
```

Comment:
Can we eliminate the pos argument here, if 'term' is a WithPos?

Reply:
As I said above, should I bound `value` with position in AST/IR?